### ISSUE TITLE
Using Funcrefs/lambdas with wiki_map_link_create

### ISSUE THREAD

**Issue by savq** [lervag/wiki.vim#104]
# Description:

Currently wiki.vim doesn't support passing funcrefs/lambdas to the `wiki_map_link_create` option (because of strange VimL behavior).

# Minimal working example

If we create a lambda: `let CreateLoudLinks = {x -> toupper(x) . '!'}`, then:

This works

```
echo exists('*CreateLoudLinks')             " string
" = 1

echo call(CreateLoudLinks, ['hello world']) " symbol
" = HELLO WORLD!
```

But this doesn't work

```
echo exists(CreateLoudLinks)                   " symbol
" = E729: using Funcref as a String

echo call('*CreateLoudLinks', ['hello world']) " string
" = E117: Unknown function: CreateLinkLinks
```

# The problem

The `exists` function always expects a string, however, the `call` function will fail if it receives a funcref as a string (but not as a symbol).

# Suggested solution

In the source code the relevant bit seems to be:

```
" wiki.vim/autoload/wiki/link/word.vim

if !empty(g:wiki_map_link_create) && exists('*' . g:wiki_map_link_create)
  let l:url = call(g:wiki_map_link_create, [a:url])
endif
```

Allowing the user to pass a funcref, and typecheck `wiki_map_link_create` here, before calling `exists` and `call` seems to me like a reasonable workaround for the inconsistency between the two functions.

**Comment by savq** [comment]
To add more context, the reason I'm trying to pass a funcref rather than the name of a function is because I'm calling a lua function, and having another function just to wrap it seems a bit cluttered. I'd rather wrap it with a lambda.

**Comment by lervag** [comment]
I agree this would be nice, but unfortunately, it is not possible. The reason is that the option names are all lowercase. Function names must start with a capital letter, see `:help E704`.

Note, this should really be just a minor inconvenience, i.e. that you need two lines instead of one, e.g.:

```vim
" We can't do this:
let g:wiki_map_link_create = {x -> toupper(x) . '!'}

" But this works
let CreateLoudLinks = {x -> toupper(x) . '!'}
let g:wiki_map_link_create = 'CreateLoudLinks'
```

**Comment by savq** [comment]
Yes, you're right. I made sure to name the lambda in the example because passing it directly wouldn't work.

On the other hand, the example you gave won't work either. If I understand correctly, when creating the lambda, it'll be stored in Vim's symbol table (`:let`), not in it's function table (`:function`); and because of this, passing the name of the lamda to `call` _as a string_ will fail, you'd have to pass it as a symbol (and it needs to start with a capital letter for that to work so we're back where we started).

Anyways, there seems to be a lot of arbitrary restrictions from vimL's part to solve this. I'm closing the issue for now, since this was more of an inquiry than an urgent bug.

Thanks for working on wiki.vim!

**Comment by lervag** [comment]
> Yes, you're right. I made sure to name the lambda in the example because passing it directly wouldn't work.
> 
> On the other hand, the example you gave won't work either. If I understand correctly, when creating the lambda, it'll be stored in Vim's symbol table (`:let`), not in it's function table (`:function`); and because of this, passing the name of the lamda to `call` _as a string_ will fail, you'd have to pass it as a symbol (and it needs to start with a capital letter for that to work so we're back where we started).
> 
> Anyways, there seems to be a lot of arbitrary restrictions from vimL's part to solve this. I'm closing the issue for now, since this was more of an inquiry than a bug.
> 
> Thanks for working on wiki.vim!

**Comment by savq** [comment]
Hey @lervag,

I decided to revisit this issue recently because it seems that since I first posted it, Neovim added a mechanism to convert Lua functions to Vim funcrefs. In practice this means that in a Lua config file one could write:

```lua
vim.g.wiki_map_link_create = function(txt) return txt:lower():gsub("%s+", "-") end
```

Lua won't complain about the capitalization of the variable, which is pretty much what I wanted in the first place.

Allowing this from wiki.vim only requires one to check if the variable is a funcref before calling `exists`.
